wai/Network/Wai.hs

```diff
@@ -315,11 +318,61 @@ ifRequest :: (Request -> Bool) -> Middleware -> Middleware
 ifRequest rpred middle app req | rpred req = middle app req
                                | otherwise =        app req
 
-
+-- $streamingRequestBodies
+--
+-- == Streaming Request Bodies
+--
+-- WAI is designed for streaming in request bodies, which allows you to process them incrementally.
+-- You can stream in the request body using functions like 'getRequestBodyChunk',
+-- the @wai-conduit@ package, or Yesod's @rawRequestBody@.
+--
+-- In the normal case, incremental processing is more efficient, since it
+-- reduces maximum total memory usage.
+-- In the worst case, it helps protect your server against denial-of-service (DOS) attacks, in which
+-- an attacker sends huge request bodies to your server.
+--
+-- Consider these tips to avoid reading the entire request body into memory:
+--
+-- * Look for library functions that support incremental processing. Sometimes these will use streaming
+-- libraries like @conduit@, @pipes@, or @streaming@.
+-- * Any attoparsec parser supports streaming input. For an example of this, see the
+-- "Data.Conduit.Attoparsec" module in @conduit-extra@.
+-- * Consider streaming directly to a file on disk. For an example of this, see the
+-- "Data.Conduit.Binary" module in @conduit-extra@.
+-- * If you need to direct the request body to multiple destinations, you can stream to both those
+-- destinations at the same time.
+-- For example, if you wanted to run an HMAC on the request body as well as parse it into JSON,
+-- you could use Conduit's @zipSinks@ to send the data to @cryptonite-conduit@'s 'sinkHMAC' and
+-- @aeson@'s Attoparsec parser.
+-- * If possible, avoid processing large data on your server at all.
+-- For example, instead of uploading a file to your server and then to AWS S3,
+-- you can have the browser upload directly to S3.
+--
+-- That said, sometimes it is convenient, or even necessary to read the whole request body into memory.
+-- For these purposes, functions like 'strictRequestBody' or 'lazyRequestBody' can be used.
+-- When this is the case, consider these strategies to mitigating potential DOS attacks:
+--
+-- * Set a limit on the request body size you allow.
```

Comment:
This makes me think we should consider taking an optional parameter `Maybe Word64` of the maximum allowed body size for the newly added functions.

Reply:
So, coming from a Yesod perspective I feel like this is already well handled by `maximumContentLength`, but I don't have a good perspective on what WAI attempts to accomplish vs Yesod/Servant/etc

Reply:
Well, that's the thing: it doesn't really try to address these things. But with this added documentation, it's begging the question: why are these functions encouraging a dangerous practice? I realize this isn't really added by your PR, since the original poorly-named functions really created the problem in the first place. But it's certainly highlighting it.

Reply:
Hm, yeah, it does highlight it. I thought about it a little more, and I don't really think having a parameter on this function to limit the body size is great from a security perspective. Thoughts:

* It would scatter the size checks across the codebase. This makes it harder to audit that it's being used properly
* It requires every developer to make security decisions they might not understand.
* In practice I would expect a lot of `Nothing` values passed (especially since frameworks like Yesod handle this in a better way, so it would be improper to try to do it here as well).

I think what Yesod doing is really great. If I added a middleware to wai-extra to limit request size (should be trivial), and pointed people to either use that or Yesod's functionality (or whatever is in their own web framework), do you think that would be sufficient? I'd rather direct people to use what I think is a great solution

Reply:
That's a great idea!